### PR TITLE
Align Tensor<rank,dim,Number> to make vectorization possible.

### PR DESCRIPTION
**Pull request by bangerth**
This is based on @kronbichler 's suggestion in https://github.com/dealii/dealii/issues/16465#issuecomment-1936351996 . This does not implement the actual use of vectorized instructions, but I think that aligning tensors by themselves is probably not a bad idea.

### CODE REVIEW

**Comment by bangerth** [comment]
This patch fails a number of tests, the simplest of which is probably `mpi_some_to_some_01.mpirun=2.debug`. I can't reproduce this here, perhaps because I have `Vectorization level:    256 bit` whereas the tester has `Vectorization level:    128 bit`. I suspect, but of course done know, that it's because we pack/unpack tensors (or in this case `Point` objects) differently if they are aligned. All of the tests that fail are MPI tests that send stuff around.

Would someone else be willing to see if this patch fails on your platform?

**Comment by kronbichler** [comment]
I checked on my machine with no architecture-specific optimization, i.e., 128 bit vectorization. I see that `mpi_some_to_some_01.mpirun=2.debug` fails here with a segmentation fault. I checked with gdb:
```
#0  0x000055555556ef90 in std::_Construct<dealii::Point<2, double>, dealii::Point<2, double> const&> (__p=0x5555559f3ea0, __args=...)
    at /usr/bin/../lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/bits/stl_construct.h:119
#1  std::__do_uninit_copy<dealii::Point<2, double> const*, dealii::Point<2, double>*> (__first=0x5555559f3ec8, __last=<optimized out>, 
    __result=<optimized out>) at /usr/bin/../lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/bits/stl_uninitialized.h:120
#2  std::__uninitialized_copy<false>::__uninit_copy<dealii::Point<2, double> const*, dealii::Point<2, double>*> (__first=<optimized out>, 
    __last=<optimized out>, __result=<optimized out>)
    at /usr/bin/../lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/bits/stl_uninitialized.h:137
#3  std::uninitialized_copy<dealii::Point<2, double> const*, dealii::Point<2, double>*> (__first=<optimized out>, __last=<optimized out>, 
    __result=<optimized out>) at /usr/bin/../lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/bits/stl_uninitialized.h:184
#4  std::__uninitialized_copy_a<dealii::Point<2, double> const*, dealii::Point<2, double>*, dealii::Point<2, double> > (
    __first=<optimized out>, __last=<optimized out>, __result=<optimized out>)
    at /usr/bin/../lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/bits/stl_uninitialized.h:373
#5  std::vector<dealii::Point<2, double>, std::allocator<dealii::Point<2, double> > >::_M_range_insert<dealii::Point<2, double> const*> (
    this=0x7fffffffcb90, __position=non-dereferenceable iterator for std::vector, __first=<optimized out>, __last=0x5555559f3ed8)
    at /usr/bin/../lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/bits/vector.tcc:814
#6  0x000055555556c5b1 in dealii::Utilities::unpack<std::vector<dealii::Point<2, double>, std::allocator<dealii::Point<2, double> > > > (
    cbegin=1 '\001', cend=33 '!', allow_compression=false) at /home/martin/deal/other/deal.II/include/deal.II/base/utilities.h:1491
#7  dealii::Utilities::unpack<std::vector<dealii::Point<2, double>, std::allocator<dealii::Point<2, double> > > > (
    buffer=std::vector of length 24, capacity 24 = {...}, allow_compression=false)
    at /home/martin/deal/other/deal.II/include/deal.II/base/utilities.h:1523
#8  dealii::Utilities::MPI::some_to_some<std::vector<dealii::Point<2, double>, std::allocator<dealii::Point<2, double> > > > (
    comm=0x7fffe29efe60 <ompi_mpi_comm_world>, objects_to_send=std::map with 1 element = {...})
    at /home/martin/deal/other/deal.II/include/deal.II/base/mpi.h:1973
#9  0x000055555556b5d1 in main (argc=1, argv=0x7fffffffd4f8) at /home/martin/deal/other/deal.II/tests/base/mpi_some_to_some_01.cc:40
```
I don't immediately see the problem, but maybe this helps. Alternatively, let me know and I can look more closely.

**Comment by bangerth** [comment]
Hm, unclear to me too. `Point<2,double>` is derived from `Tensor<1,2,double>`. With 128-bit vectorization, we have `VectorizedArray<double,2>` available, which should be aligned to a 16-byte boundary, right? In that case, `Tensor<1,2,double>` (and consequently `Point<2,double>` also needs to be aligned to a 16-byte boundary and that's what the `alignas` part of this patch does. The way I read the backtrace is that it tries to construct the object at 0x5555559f3ea0, which is indeed a 16-byte aligned address. But in frame 1, it looks like (though I'm not sure what the arguments exactly mean) that it's trying to *read* a `Point<2,double>` from address 0x5555559f3ec8, which is not aligned.

I *think* that a frame is missing in the backtrace, namely to this function called in line 1491 of `utilities.h`:
```
    template <typename T,
              typename = std::enable_if_t<!std::is_same_v<T, bool> &&
                                          std::is_trivially_copyable_v<T>>>
    inline void
    create_vector_of_trivially_copyable_from_buffer(
      const std::vector<char>::const_iterator &cbegin,
      const std::vector<char>::const_iterator &cend,
      std::vector<T>                          &object)
    {
      // The size of the object vector can be found in cbegin of the buffer.
      // The data starts at cbegin + sizeof(vector_size).

      // Get the size of the vector
      typename std::vector<T>::size_type vector_size;
      memcpy(&vector_size, &*cbegin, sizeof(vector_size));

      Assert(static_cast<std::ptrdiff_t>(cend - cbegin) ==
               static_cast<std::ptrdiff_t>(sizeof(vector_size) +
                                           vector_size * sizeof(T)),
             ExcMessage("The given buffer has the wrong size."));
      (void)cend;

      // Copy the elements:
      object.clear();
      if (vector_size > 0)
        {
          const T *const buffer_data_begin =
            reinterpret_cast<const T *>(&*cbegin + sizeof(vector_size));
          const T *const buffer_data_end = buffer_data_begin + vector_size;
          object.insert(object.end(), buffer_data_begin, buffer_data_end);
        }
    }
```
I suspect it's the last line that goes wrong where we copy the object *as that object*. We need to copy just the bits, which we can because it is trivially copyable. If you agree that that seems like the right analysis, I'll prepare a patch later today or tomorrow.

In any case, thank you for the backtrace!

File: include/deal.II/base/utilities.h [hunk 1]
```diff
@@ -1295,15 +1295,29 @@ namespace Utilities
...
-      // Copy the elements:
-      object.clear();
+      // Resize the output array and copy the elements into it. We need
+      // to make sure that we don't access the buffer via a
+      // reinterpret_cast<T*> because the T objects in the buffer may
+      // not be aligned properly. Rather, use memcpy, which doesn't care
+      // about alignment and is correct in this situation because we
+      // are dealing with trivially copyable objects.
+      //
+      // (Strictly speaking, this write into the output object twice, once
```
Comment (masterleinad) [inline include/deal.II/base/utilities.h:1305]:
```suggestion
      // (Strictly speaking, this writes into the output object twice, once
```

**Comment by bangerth** [comment]
OK, that took unnecessarily many iterations, but it seems to work now...

**Review by kronbichler — APPROVED** [review]
I like this version.

**Pull request merged**